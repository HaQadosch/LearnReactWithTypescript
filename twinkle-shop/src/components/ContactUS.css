.form {
  width: 300px;
  margin: 0px auto 0px auto;
}

.form-group {
  display: flex;
  flex-direction: column;
  margin-bottom: 20px;
}

.form-group label {
  align-self: flex-start;
  font-size: 16px;
  margin-bottom: 3px;
}

.form-group input, select, textarea {
  font-family: Arial;
  font-size: 16px;
  padding: 5px;
  border: lightgray solid 1px;
  border-radius: 5px;
}