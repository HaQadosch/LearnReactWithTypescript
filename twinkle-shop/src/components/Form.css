.form-error {
  font-size: 13px;
  color: red;
  margin: 3px auto 0px 0px;
}

.form-group button {
  font-size: 16px;
  padding: 8px 5px;
  width: 80px;
  border: black solid 1px;
  border-radius: 5px;
  background-color: black;
  color: white;
}
.form-group button:disabled {
  border: gray solid 1px;
  background-color: gray;
  cursor: not-allowed;
}