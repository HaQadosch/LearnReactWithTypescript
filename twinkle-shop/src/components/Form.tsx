import React, { useState, createContext } from 'react'
import produce from 'immer'

interface IValues {
  [key: string]: any

}

interface IFormProps {
  defaultValues: IValues
}

interface IFormFieldProps {
  name: string
  label: string
  type?: 'Text' | 'Email' | 'Select' | 'TextArea'
  options?: string[]
}

interface IFormContext {
  values: IValues
  setValue?: (fieldname: string, value: any) => void
}

const FormContext = createContext<IFormContext>({ values: {} })

export const FormField: React.FC<IFormFieldProps> = ({ name, label, type = 'Text', options }) => {
  return (
    <FormContext.Consumer>
      {({ values, setValue }) => {
        const handleChange: React.ChangeEventHandler<HTMLTextAreaElement | HTMLSelectElement | HTMLInputElement> = ({ currentTarget: { value: targetValue } }) => {
          if (setValue) {
            setValue(name, targetValue)
          }
        }
        return <fieldset className='form-group'>
          <label htmlFor={name}>{label}</label>
          {(type === 'Text' || type === 'Email') && <input type={type.toLowerCase()} name={name} id={name} value={values[name]} onChange={handleChange} />}
          {(type === 'TextArea') && <textarea id={name} value={values[name]} onChange={handleChange} />}
          {(type === 'Select') && (
            <select name={name} id={name} value={values[name]} onChange={handleChange} >
              {options && options.map((opt, ind) => <option value={opt} key={ind} >{opt}</option>)}
            </select>
          )}
        </fieldset>
      }
      }
    </FormContext.Consumer>
  )
}

export const Form: React.FC<IFormProps> = ({ defaultValues, children }) => {
  const [values, setValues] = useState<IValues>(defaultValues)

  const setNewValues = (fieldname: string, value: any) => {
    setValues(produce(values, draft => {
      draft[fieldname] = value
    }))
  }

  const handleFormSubmit: React.ChangeEventHandler<HTMLFormElement> = evt => {
    evt.preventDefault()
  }

  const context: IFormContext = { values, setValue: setNewValues }

  return (
    <FormContext.Provider value={context}>
      <form className='form' noValidate={true} onSubmit={handleFormSubmit} >
        {children}
      </form>
    </FormContext.Provider>
  )
}
