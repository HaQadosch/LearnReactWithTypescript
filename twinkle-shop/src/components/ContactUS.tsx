import React from 'react'
import './ContactUS.css'
import { Form, FormField, required, minLength, ISubmitResult, IValues } from './Form';
import { wait } from '../ProductData'
interface IContactUsProps {
  onSubmit: (values: IValues) => Promise<ISubmitResult>
}

export const ContactUS: React.FC<IContactUsProps> = ({ onSubmit }) => {
  const handleSubmit = async (values: IValues): Promise<ISubmitResult> => {
    const result = await onSubmit(values)
    return result
  }

  return (
    <Form data-testid='contact-us'
      defaultValues={{ name: '', email: '', reason: 'Support', notes: '' }}
      validationRules={{
        email: { validator: required },
        name: [{ validator: required }, { validator: minLength, arg: 2 }]
      }}
      onSubmit={handleSubmit}
    >
      <FormField name='name' label='Your name' />
      <FormField name='email' label='Your email' type='Email' />
      <FormField name='reason' label='Reason you need to contact us' type='Select' options={['Marketing', 'Support', 'Feedback', 'Jobs', 'Offer']} />
      <FormField name='notes' label='Additional Notes' type='TextArea' />
    </Form>
  )
}

export const ContactUSPage: React.FC = () => {
  const handleSubmit = async (values: IValues): Promise<ISubmitResult> => {
    await wait(1000)
    return {
      errors: {
        email: ['Something is wrong with this.']
      },
      success: false
    }
  }

  return (
    <section className='page-container'>
      <h2>Contact US</h2>
      <p>If you enter your details, we'll get back to you as soon as possible.</p>
      <ContactUS onSubmit={handleSubmit} />
    </section>
  )
}
