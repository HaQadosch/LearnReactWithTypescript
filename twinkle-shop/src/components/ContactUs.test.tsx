import React from 'react'
// import ReactDOM from 'react-dom'
import { render, fireEvent, cleanup } from '@testing-library/react'
import '../setupTests'

import { ContactUS } from './ContactUS'
import { ISubmitResult } from './Form'

const handleSubmit = jest.fn<Promise<ISubmitResult>, any[]>(async () => {
  return { success: true }
})

afterEach(cleanup)

describe('<ContactUS />', () => {
  it('should display 2 error messages when submitted empty.', () => {
    const { getByText, getAllByText } = render(<ContactUS onSubmit={handleSubmit} />)

    const submitBtn = getByText('Submit')

    fireEvent.click(submitBtn)

    expect(handleSubmit).toHaveBeenCalledTimes(1)
    expect(handleSubmit).toHaveBeenCalledWith({"email": "", "name": "", "notes": "", "reason": "Support"})
    expect(getAllByText('This must be filled in.')).toHaveLength(2)
  });
  
  it('should submit okay if filled in properly', () => {
    const { getByText, getByLabelText, queryByText, debug } = render(<ContactUS onSubmit={handleSubmit} />)

    const nameInput = getByLabelText('Your name')
    const emailInput = getByLabelText('Your email')
    const submitBtn = getByText('Submit')

    fireEvent.change(nameInput, { target: { value: 'Carl'}})
    fireEvent.change(emailInput, { target: { value: 'carl@rippon.eu'}})
    fireEvent.click(submitBtn)

    expect(queryByText('This must be filled in.')).not.toBeInTheDocument()
  });
});
