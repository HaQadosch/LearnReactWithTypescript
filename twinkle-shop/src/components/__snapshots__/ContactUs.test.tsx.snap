// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactUS /> should renders ok 1`] = `
<div>
  <form
    class="form"
    data-testid="form"
    novalidate=""
  >
    <fieldset
      class="form-group"
    >
      <label
        for="name"
      >
        Your name
      </label>
      <input
        id="name"
        name="name"
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      class="form-group"
    >
      <label
        for="email"
      >
        Your email
      </label>
      <input
        id="email"
        name="email"
        type="email"
        value=""
      />
    </fieldset>
    <fieldset
      class="form-group"
    >
      <label
        for="reason"
      >
        Reason you need to contact us
      </label>
      <select
        id="reason"
        name="reason"
      >
        <option
          value="Marketing"
        >
          Marketing
        </option>
        <option
          value="Support"
        >
          Support
        </option>
        <option
          value="Feedback"
        >
          Feedback
        </option>
        <option
          value="Jobs"
        >
          Jobs
        </option>
        <option
          value="Offer"
        >
          Offer
        </option>
      </select>
    </fieldset>
    <fieldset
      class="form-group"
    >
      <label
        for="notes"
      >
        Additional Notes
      </label>
      <textarea
        id="notes"
      />
    </fieldset>
    <div
      class="form-group"
    >
      <button
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
